### PR TITLE
Teach info bars to be dismissed permanently

### LINKED ISSUE #10798: Make exit code behavior more disacoverable
# Description of the new feature/enhancement
When a process exits with a return code while closing a terminal, the following message shows up (assuming exit 130)
```
[process exited with code 130]
```

It's difficult to know what this means, especially if it's while closing a pane.  If a process returns an exit code that is non zero, I can't figure out how to get rid of that pane now.  Example:
image

This inevitably leads eventually to https://github.com/microsoft/terminal/issues/5214

# Proposed technical implementation details (optional)
Add a clearer message why this happened, and how to change it.
```
Process exited with code 130, to adjust what happens with non-zero exit codes visit Settings->Foo->bar
```

Then in the GUI add a new option that will control adding
```
"closeOnExit": "always"
```

### PR DESCRIPTION
**Pull request by Don-Vito**
## Summary of the Pull Request
* Introduces info bar shown upon session failure, 
that guides the user how to configure termination behavior
  * Allows this info bar to be dismissed permanently (choice stored in state) 
* Allows "keyboard service" info bar to be dismissed permanently

## PR Checklist
* [x] Closes #10798, #8699
* [x] CLA signed. 
* [ ] Tests added/passed
* [ ] Documentation updated. 
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already.

## Detailed Description of the Pull Request / Additional comments
UI:
* Introduce an additional info bar for "close on exit" configuration tip
  * Stack this bar after "keyboard service" bar
* Add "Don't show again" button to both bars

Dismiss Permanently:
* Introduce a set of "dismissed messages" to the Application State
* Add verification the message is not dismissed before showing an info bar
* "Don't show  again" persists the choice under "dismissed messages"

Wiring the Info Bar:
* Register `TerminalPage` on `TermControl`'s `ConnectionStateChanged` event
* Once event is triggered check whether the state is failure
* If so and the message was not dismissed permanently, show the info bar

### CODE REVIEW

**Comment by Don-Vito** [comment]
image

File: src/cascadia/TerminalApp/Resources/en-US/Resources.resw
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Resources/en-US/Resources.resw:708]: I'd prefer that we make this one resource and bind to the one resource. That'll reduce the localization burden and make it easier for us to add new "Don't show again" buttons in the future.

File: src/cascadia/TerminalApp/Resources/en-US/Resources.resw
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Resources/en-US/Resources.resw:702]: Idea: I think it'd be cool if (later) we made "advanced profile settings" a hyperlink that opened the settings UI for this profile and navigated to the relevant page.

The only reason I bring that up now is that if we want to pursue this, we should maybe break up this resource into two:
- "Termination behavior can be configured in {}"
- "advanced profile settings"
And in TerminalPage.cpp, we replace `{}` with "advanced profile settings" and (later) hyperlink-ify it.

Thoughts?

File: src/cascadia/TerminalSettingsModel/EnumMappings.idl
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsModel/EnumMappings.idl:29]: I don't think we need any of the EnumMappings changes. We use these to expose the `JSON_ENUM_MAPPER` from TerminalSettingsSerializationHelpers.h over to the settings UI.

Since we don't want to put these in the settings UI and `ApplicationState` is already accessible via the settings model, no need.

File: src/cascadia/TerminalApp/Resources/en-US/Resources.resw
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/Resources/en-US/Resources.resw:701]:
```suggestion
    <value>Termination behavior can be configured in advanced profile settings.</value>
```
for consistency

**Review by carlos-zamora — CHANGES_REQUESTED** [review]
Just a few small things before I approve

**Comment by zadjii-msft** [comment]
Okay I need top put aside more time for a full review, but this is legit.

\<crazy idea hat>

How hard would it be to add a link to the docs (like, to https://docs.microsoft.com/en-us/windows/terminal/customize-settings/profile-advanced#profile-termination-behavior) to that infobar? We could generate the fwlink if needed

File: src/cascadia/TerminalApp/Resources/en-US/Resources.resw
Comment (DHowett) [reply src/cascadia/TerminalApp/Resources/en-US/Resources.resw:708]: I wish we didn't have to do that. But also, having multiple resources that say the same thing is not really that bad.

File: src/cascadia/TerminalApp/Resources/en-US/Resources.resw
Comment (DHowett) [reply src/cascadia/TerminalApp/Resources/en-US/Resources.resw:701]: We'll want to reword generally anyway

File: src/cascadia/TerminalApp/Resources/en-US/Resources.resw
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Resources/en-US/Resources.resw:702]: I'm gonna say this is something that can be done in a follow up

**Review by zadjii-msft — APPROVED** [review]
Besides wording nits, I'm cool with the rest of the code here. I'll leave it to the rest of the team to sort out the specific wording (/cc @cinnamon-msft)

File: src/cascadia/TerminalApp/Resources/en-US/Resources.resw
Comment (Don-Vito) [reply src/cascadia/TerminalApp/Resources/en-US/Resources.resw]: Will leave this as a follow up as it really depends on what UX we implement (i.e., probably we will need to put the hyperlink as a separate element)

**Comment by Don-Vito** [comment]
> How hard would it be to add a link to the docs (like, to https://docs.microsoft.com/en-us/windows/terminal/customize-settings/profile-advanced#profile-termination-behavior) to that infobar? We could generate the fwlink if needed

This is not very hard. We can do something like

image

 but we need to decide if we want to follow up by opening the settings :blush:.... or... we can do all of them:
image

In any case suggest to do it in a follow up. I can open a ticket for it.. and proceed with it if the team is interested.

**Comment by cinnamon-msft** [comment]
I am a fan of this. :)

**Review by cinnamon-msft — APPROVED** [review]

**Review by carlos-zamora — APPROVED** [review]
Perfect. Let's do this

**Pull request merged**
